hydradx/model/amm/omnipool_amm.py

```diff
@@ -86,63 +146,120 @@ def swap_lrna(
+#         old_state: dict,
+#         old_agents: dict,
+#         trader_id: string,
+#         delta_R: float,
+#         delta_Q: float,
+#         i: int,
+#         fee_assets: float = 0,
+#         fee_lrna: float = 0
+# ) -> tuple:
+#     """Computed new state for LRNA swap with fee"""
+#     new_state, new_agents = swap_lrna(old_state, old_agents, trader_id, delta_R, delta_Q, i)
+#     delta_Q = new_state['Q'][i] - old_state['Q'][i]
+#     if delta_Q < 0:
+#         # LRNA fee
+#         new_agents[trader_id]['q'] += delta_Q * fee_lrna
+#         new_state['D'] -= delta_Q * fee_lrna
+#     else:
+#         delta_R = new_state['R'][i] - old_state['R'][i]  # delta_R is negative
+#         # asset fee
+#         new_agents[trader_id]['r'][i] += delta_R * fee_assets
+#         # fee added back as liquidity, distributed to existing LPs (i.e. no new shares minted)
+#         # eventually, can mint protocol shares to take some cut as POL
+#         p = price_i(new_state, i)
+#         new_state['R'][i] -= delta_R * fee_assets
+#         # LRNA minted so that asset fee level does not change price and increase IL
+#         new_state['Q'][i] = p * new_state['R'][i]
+#     return new_state, new_agents
+
+
+def swap_assets_direct(
         old_state: dict,
         old_agents: dict,
         trader_id: string,
-        delta_R: float,
-        delta_Q: float,
-        i: int,
+        delta_token: float,
+        i_buy: int,
+        i_sell: int,
         fee_assets: float = 0,
         fee_lrna: float = 0
 ) -> tuple:
-    """Computed new state for LRNA swap with fee"""
-    new_state, new_agents = swap_lrna(old_state, old_agents, trader_id, delta_R, delta_Q, i)
-    delta_Q = new_state['Q'][i] - old_state['Q'][i]
-    if delta_Q < 0:
-        # LRNA fee
-        new_agents[trader_id]['q'] += delta_Q * fee_lrna
-        new_state['D'] -= delta_Q * fee_lrna
-    else:
-        delta_R = new_state['R'][i] - old_state['R'][i]  # delta_R is negative
-        # asset fee
-        new_agents[trader_id]['r'][i] += delta_R * fee_assets
-        # fee added back as liquidity, distributed to existing LPs (i.e. no new shares minted)
-        # eventually, can mint protocol shares to take some cut as POL
-        p = price_i(new_state, i)
-        new_state['R'][i] -= delta_R * fee_assets
-        # LRNA minted so that asset fee level does not change price and increase IL
-        new_state['Q'][i] = p * new_state['R'][i]
+    i = i_sell
+    j = i_buy
+    delta_Ri = delta_token
+    assert delta_Ri > 0, 'sell amount must be greater than zero'
+
+    delta_Qi = old_state['Q'][i] * -delta_Ri / (old_state['R'][i] + delta_Ri)
+    delta_Qj = -delta_Qi * (1 - fee_lrna)
+    delta_Rj = old_state['R'][j] * -delta_Qj / (old_state['Q'][j] + delta_Qj) * (1 - fee_assets)
+    delta_L = min(-delta_Qi * fee_lrna, -old_state['L'])
+    delta_QH = -fee_lrna * delta_Qi - delta_L
+
+    new_state = copy.deepcopy(old_state)
+    new_state['Q'][i] += delta_Qi
```

Comment:
just wondering if here it should add delta_QH  since the part of it is going to LRNA side of hdx subpool.

I mean something like this: 

`new_state['Q'][i] += delta_Qi - delta_QH`

Reply:
I don't think so. The fee is subtracted earlier from delta_Qj and then it goes to the hdx subpool here:
    new_state['Q'][new_state['token_list'].index('HDX')] += delta_QH
on line 260